In [2]:
import numpy as np
import random
import math
import matplotlib.pyplot as plt


**1. Programmer  l'algorithme du perceptron  a)  version batch et  b)  version
online.**

In [42]:
# version batch 
def versionBatch(W,alph,X,etiq):
  itera_bch=0
  pred=np.ones(len(etiq))#mon tableau d'étiquettes que je vais prédir
  n=len(X)
  W = np.append(1, W)#ajouter le biais à mes poids
  #print(W)
  while not classeCorrectementVecteur(etiq, pred): #tanque les 2 liste ne sont pas pareille je continue
    Dw=0
    for i in range(n):
      itera_bch+=1
      x_ = np.append(1, X[i])
      if np.transpose(W).dot(x_)> 0 :
        pred[i]=1
      else:
        pred[i]=-1
      if pred[i] != etiq[i] :
        Dw = x_.dot(alph*(etiq[i]-pred[i]))
      W=W+Dw
        
    
  return W,pred,itera_bch

In [43]:
#version online
def versionOnline(W,alph,X,etiq):
  pred=np.ones(len(X))
  W = np.append(1, W)
  IT=0
  while not classeCorrectementVecteur(etiq, pred):
    i=random.randint(0, len(X) - 1)#car les vecteurs commence de 0 à n-1
    x_ = np.append(1, X[i])
    IT+=1
    Dw=0
    if np.transpose(W).dot(x_)> 0 :
       pred[i]=1
    else:
       pred[i]=-1
    if pred[i] != etiq[i] :
      Dw = x_.dot(alph*(etiq[i]-pred[i]))
    W=W+Dw
      
    
  return W,pred,IT
  

    
  

In [44]:
def classeCorrectementVecteur(vect1,vect2):
  
  
  if np.array_equal(vect1,vect2):
    
      return True 
  return False
    

Tester la version batch ET / OU / ExempleCours

In [45]:
W=np.array([0.4,0.6])
alph=0.8


X_ = np.array([[2,1],[0,-1],[-2,1],[0,2]])
etiq = np.array([1,1,-1,-1])
print(versionBatch(W,alph,X_,etiq))


X_ou= np.array([[0,0],[1,0],[0,1],[1,1]]) 
etiq_ou = np.array([-1,1,1,1])
print(versionBatch(W,alph,X_ou,etiq_ou))

X_et= np.array([[0,0],[1,0],[0,1],[1,1]])
etiq_et = np.array([-1,-1,-1,1])


print(versionBatch(W,alph,X_et,etiq_et))

(array([-2.2,  6.8, -2.6]), array([ 1.,  1., -1., -1.]), 8)
(array([-0.6,  6.8,  7. ]), array([-1.,  1.,  1.,  1.]), 24)
(array([-2.2,  2. ,  2.2]), array([-1., -1., -1.,  1.]), 8)


Tester la version Online ET / OU / ExempleCours

In [46]:
print(versionOnline(W,alph,X_,etiq))
print(versionOnline(W,alph,X_ou,etiq_ou))
print(versionOnline(W,alph,X_et,etiq_et))

(array([-0.6,  3.6, -1. ]), array([ 1.,  1., -1., -1.]), 13)
(array([-0.6,  2. ,  0.6]), array([-1.,  1.,  1.,  1.]), 13)
(array([-0.6,  0.4,  0.6]), array([-1., -1., -1.,  1.]), 6)


**2. Génerer Données  LS  aléatoires.  Construire  un  ensemble  LS  de  P  exemples  en  N+1
dimensions avec un p  erceptron p  rofesseur**  

In [47]:
# Génération de l'ensemble des données pour le Perceptron Professeur
def genererDataset(P, N):
    #si je n'effectue pas la normalisation mes algo ne s'arrete quasiment pas car beaucoup de valeurs après la virgule
    X = np.random.randint(-100,100,(P,N))/100 
    W = np.random.randint(-100,100,N)/100
    
    
    return X,  W
    
genererDataset(100, 10)

(array([[ 0.7 ,  0.8 ,  0.95, -0.86,  0.17,  0.39, -0.78,  0.66,  0.15,
          0.32],
        [-0.37, -0.3 ,  0.42,  0.89, -0.09,  0.71, -0.93, -0.4 ,  0.42,
          0.11],
        [ 0.06, -0.54,  0.36,  0.24,  0.72, -0.35, -0.83,  0.13, -0.91,
         -0.32],
        [-0.41, -0.23,  0.52, -0.39, -0.71, -0.66,  0.39, -0.76, -0.06,
          0.38],
        [ 0.18,  0.  , -0.26, -0.86,  0.31, -0.59, -0.59, -0.63,  0.33,
         -0.81],
        [ 0.92,  0.44,  0.63,  0.28,  0.2 ,  0.42,  0.02,  0.96,  0.52,
          0.26],
        [-0.47, -0.1 , -0.99, -0.04, -0.75,  0.61,  0.15,  0.33,  0.84,
         -0.59],
        [-0.91,  0.46, -0.14,  0.85, -0.57, -0.2 , -0.54,  0.7 ,  0.97,
         -0.3 ],
        [-0.53, -0.13,  0.75, -0.12,  0.8 , -0.72, -0.13,  0.6 ,  0.02,
         -0.2 ],
        [-0.72, -0.12,  0.76, -0.77,  0.02,  0.14,  0.29, -0.22, -0.11,
         -0.4 ],
        [ 0.75,  0.05, -0.63, -0.65,  0.3 ,  0.57, -0.61, -0.31,  0.88,
         -0.34],
        [-0.13, -0.91

In [48]:
def Linear_separability(X,w):
  tau=[]
  for x in X:
       
       if np.dot(w,x)<=0:
           tau += [-1]
           #plt.scatter(X[i,0],X[i,1],c="r")
       else:
           tau += [1]
           #plt.scatter(X[i,0],X[i,1],c="g")
       
  #plt.show
  return tau


**3. Apprentissage. Apprendre avec a) la version batch de l'algorithme du perceptron et
b) la  version online.  Pour chaque version faire apprendre un  perceptron élève W sur
l'ensemble LS obtenu lors de la génération de données.**

R = cos(alpha) = a * b / (|a| * |b|)

Lancer  le  programme  avec  les  valeurs  :  N=2,10,100,500,1000 ;
P=10,100,500,1000. Donner les résultats sous la forme de 2 tableaux (batch et online) où
chaque case affiche la moyenne du nombre d'itérations <IT> et la moyenne du rapport
<R>, sur 50 tirages aléatoires

In [31]:
def ApprentissageEleveOnline(P, N):
    moyenneIT=0
    moyenneR= 0  
    for i in range(50): #50 tirage aléatoires
      X_prof,W_prof=genererDataset(P, N)

      etiq_prof= Linear_separability(X_prof,W_prof)
      W_eleve,pred_eleve,IT = versionOnline(W_prof,alph,X_prof,etiq_prof)
      moyenneIT=moyenneIT+IT
      W_prof = np.append(1, W_prof)
      #print("etiquette professeur",etiq_prof)
      #print("etiquette professeur",pred_eleve)
      print("\n")
      R = math.cos(np.dot(W_prof, W_eleve) / np.dot(np.linalg.norm(W_prof), np.linalg.norm(W_eleve)))
      moyenneR= moyenneR+R
    moyenneIT=moyenneIT/50
    moyenneR= moyenneR/50


    return  moyenneIT,moyenneR



#############
moyenneIT ,moyenneR= ApprentissageEleveOnline(10, 2)
print(moyenneIT)
print(moyenneR)





































































































47.94
0.8134888164574082


version online :



**P=10**

p=10 N=2 moyIT => 51.74 moyR=>0.84

p=10 N=10 moyIT => 44.04  moyR=>0.846

p=10 N=100 moyIT =>32.5 moyR=>0.8106

p=10 N=500 moyIT =>23.28  moyR=> 0.69

p=10 N=1000 moyIT =>31.3  moyR=>0.715


**P=100**

p=100 N=2 moyIT => 2754.22 moyR=>0.813

p=100 N=10 moyIT => 2045.64 moyR=>0.663

p=100 N=100 moyIT =>903.02 moyR=>0.79

p=100 N=500 moyIT =>  735.9  moyR=>0.93

p=100 N=1000 moyIT => 795.24 moyR=>0.96


**P=500**

p=500 N=2 moyIT =>82970.16 moyR=> 0.82

p=500 N=10 moyIT =>  87868.28 moyR=> 0.68

p=500 N=100 moyIT =>22065.9  moyR=> 0.587

p=500 N=500 moyIT =>8179.02 moyR=> 0.793


p=500 N=1000 moyIT => 6194.56 moyR=> 0.88


**P=1000**

p=1000 N=2 moyIT => 392813.04 moyR=>0.834

p=1000 N=10 moyIT => 400228.08 moyR=> 0.65

p=1000 N=100 moyIT => 117424.94 moyR=> 0.563

p=1000 N=500 moyIT =>27000.04 moyR=>0.680

p=1000 N=1000 moyIT => 18724.86  moyR=>0.792





In [40]:
def ApprentissageEleveBatch(P, N):
    moyenneIT=0
    moyenneR= 0
    for i in range(50):#50 tirages
      X_prof,W_prof=genererDataset(P, N)
      etiq_prof= Linear_separability(X_prof,W_prof)
      
      W_eleve,pred_eleve,IT = versionBatch(W_prof,alph,X_prof,etiq_prof)
      moyenneIT=moyenneIT+IT
      W_prof = np.append(1, W_prof)
      print("etiquette professeur",etiq_prof)
      print("etiquette professeur",pred_eleve)
      print("\n")
      R = math.cos(np.dot(W_prof, W_eleve) / np.dot(np.linalg.norm(W_prof), np.linalg.norm(W_eleve)))
      moyenneR= moyenneR+R
    moyenneIT=moyenneIT/50
    moyenneR= moyenneR/50


    return  moyenneIT,moyenneR



##########
moyenneIT ,moyenneR= ApprentissageEleveBatch(1000, 2)

print(moyenneIT)
print(moyenneR)

etiquette professeur [-1, 1, -1, 1, 1, -1, 1, 1, -1, -1, 1, 1, 1, 1, 1, 1, -1, 1, -1, 1, 1, -1, 1, -1, 1, 1, -1, -1, -1, -1, 1, -1, -1, 1, -1, -1, 1, 1, -1, -1, -1, -1, -1, 1, -1, 1, 1, 1, -1, -1, -1, 1, 1, -1, 1, -1, -1, -1, -1, 1, 1, 1, 1, 1, 1, 1, -1, 1, -1, -1, -1, 1, 1, 1, 1, 1, -1, -1, 1, 1, 1, -1, 1, -1, 1, 1, -1, 1, 1, 1, -1, 1, -1, 1, 1, 1, 1, 1, -1, -1, 1, 1, 1, -1, 1, -1, 1, 1, -1, 1, -1, -1, -1, -1, 1, -1, -1, -1, -1, -1, 1, -1, -1, 1, -1, 1, 1, 1, 1, -1, 1, -1, -1, -1, -1, 1, 1, 1, -1, -1, -1, -1, -1, 1, -1, 1, 1, 1, -1, 1, -1, 1, 1, 1, -1, 1, -1, 1, 1, -1, -1, -1, 1, -1, -1, -1, 1, 1, -1, -1, 1, 1, 1, -1, 1, -1, -1, -1, -1, -1, 1, -1, -1, -1, 1, 1, -1, 1, -1, -1, -1, -1, 1, -1, -1, -1, -1, 1, -1, -1, 1, 1, -1, 1, 1, -1, -1, -1, 1, -1, -1, 1, 1, -1, 1, 1, -1, 1, 1, 1, -1, -1, 1, -1, 1, -1, 1, 1, 1, -1, -1, 1, -1, 1, -1, -1, 1, -1, 1, 1, -1, -1, -1, -1, -1, 1, -1, 1, -1, -1, -1, 1, 1, -1, -1, 1, -1, -1, 1, -1, 1, -1, -1, 1, -1, 1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, 1, -

Version Batch

P=10

p=10 N=2 moyIT =>91.4  moyR=>0.83

p=10 N=10 moyIT =>45.6  moyR=>0.849

p=10 N=100 moyIT =>29.4 moyR=>0.83

p=10 N=500 moyIT =>22.0 moyR=>0.738

p=10 N=1000 moyIT =>19.2 moyR=>0.694


P=100

p=100 N=2 moyIT =>9346.0  moyR=>0.828

p=100 N=10 moyIT => 6126.0 moyR=>0.664

p=100 N=100 moyIT =>2556.0 moyR=>0.817

p=100 N=500 moyIT => 1658.0 moyR=>0.966

p=100 N=1000 moyIT =>1334.0  moyR=>0.98


**P=500**

p=500 N=2 moyIT =>588280.0 moyR=>0.83

p=500 N=10 moyIT => 353500.0   moyR=> 0.64

p=500 N=100 moyIT => 121370.0 moyR=>0.589

p=500 N=500 moyIT => 46660.0 moyR=> 0.81

p=500 N=1000 moyIT => 36770.0 moyR=>0.91



**P=1000**

p=1000 N=2 moyIT => 827020.0 moyR=>0.82

p=1000 N=10 moyIT =>  2149060.0 moyR=>0.644

p=1000 N=100 moyIT =>734980.0 moyR=>0.56

p=1000 N=500 moyIT => 232740.0 moyR=>0.69

p=1000 N=1000 moyIT => 172820.0 moyR=>0.81